pkg/controller/job/backoff_utils.go

```diff
@@ -0,0 +1,189 @@
+		backoffInfo.lastFailureTime = lastFailureTime
+	}
+}
+
+func (s *backoffInfoStore) removeBackoffInfo(jobId string) {
+	if b, exists, _ := s.store.GetByKey(jobId); exists {
+		s.store.Delete(b)
+	}
+}
+
+func newBackoffInfoStore() *backoffInfoStore {
+	return &backoffInfoStore{
+		store: cache.NewStore(backoffInfoKeyFunc),
+	}
+}
+
+var backoffInfoKeyFunc = func(obj interface{}) (string, error) {
+	if u, ok := obj.(*backoffInfo); ok {
+		return u.key, nil
+	}
+	return "", fmt.Errorf("could not find key for obj %#v", obj)
+}
+
+func getNewBackoffInfo(clock clock.WithTicker, backoffInfoStore *backoffInfoStore, key string, newSucceededPods []*v1.Pod, newFailedPods []*v1.Pod) backoffInfo {
+	currentTime := metav1.NewTime(clock.Now())
+	var oldBackoffInfo *backoffInfo
+
+	if b, exists, _ := backoffInfoStore.store.GetByKey(key); exists {
+		oldBackoffInfo = b.(*backoffInfo)
+	} else {
+		oldBackoffInfo = &backoffInfo{
+			key:                      key,
+			failuresAfterLastSuccess: 0,
+			lastFailureTime:          nil,
+		}
+	}
+
+	sortByFinishedTime(newSucceededPods, currentTime)
+	sortByFinishedTime(newFailedPods, currentTime)
+	newBackoffInfo := backoffInfo{
+		key,
+		oldBackoffInfo.failuresAfterLastSuccess,
+		oldBackoffInfo.lastFailureTime,
+	}
+
+	if len(newSucceededPods) == 0 && len(newFailedPods) == 0 {
+		return newBackoffInfo
+	}
+
+	if len(newFailedPods) == 0 {
+		newBackoffInfo.failuresAfterLastSuccess = 0
+		newBackoffInfo.lastFailureTime = nil
+		return newBackoffInfo
+	}
+
+	if len(newSucceededPods) == 0 {
+		newBackoffInfo.failuresAfterLastSuccess = newBackoffInfo.failuresAfterLastSuccess + int32(len(newFailedPods))
+		lastFailureTime := getFinishedTime(newFailedPods[len(newFailedPods)-1], currentTime)
+		newBackoffInfo.lastFailureTime = &lastFailureTime
+		return newBackoffInfo
+	}
+
+	newBackoffInfo.failuresAfterLastSuccess = 0
+	newBackoffInfo.lastFailureTime = nil
+
+	lastSuccessTime := getFinishedTime(newSucceededPods[len(newSucceededPods)-1], currentTime)
+	for i := len(newFailedPods) - 1; i >= 0; i-- {
+		failedTime := getFinishedTime(newFailedPods[i], currentTime)
+		if lastSuccessTime.After(failedTime.Time) || lastSuccessTime.Equal(&failedTime) {
+			break
+		}
+		if newBackoffInfo.lastFailureTime == nil {
+			newBackoffInfo.lastFailureTime = &failedTime
+		}
+		newBackoffInfo.failuresAfterLastSuccess = newBackoffInfo.failuresAfterLastSuccess + 1
+	}
+
+	return newBackoffInfo
+}
+
+func sortByFinishedTime(pods []*v1.Pod, currentTime metav1.Time) {
+	finishedTimeSortFunction := func(p1, p2 *v1.Pod) bool {
+		p1FinishTime := getFinishedTime(p1, currentTime)
+		p2FinishTime := getFinishedTime(p2, currentTime)
+
+		return p1FinishTime.Before(&p2FinishTime)
+	}
+
+	sort.Slice(pods, func(i, j int) bool {
+		return finishedTimeSortFunction(pods[i], pods[j])
+	})
+}
+
+func getFinishedTime(p *v1.Pod, currentTime metav1.Time) metav1.Time {
+	var finishTime *metav1.Time
```

Comment:
```suggestion
	var finishTime time.Time
```

which is the oldest possible date

Reply:
Done